samwise.js

```diff
@@ -1,23 +1,21 @@
-/**
```

Comment:
SCRIPT-This function seems to have grown quite complex. Can we break it down further?